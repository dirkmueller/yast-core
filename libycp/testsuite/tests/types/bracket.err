Parsed:
----------------------------------------------------------------------
{
    // list a
    // filename: "tests/types/bracket.ycp"
    list a = ["foo"];
    return (a[0]:nil == "foo");
}
----------------------------------------------------------------------
[Parser] tests/types/bracket.ycp:12 Bracket parameter '"bar"' to 'a' has wrong type
[Parser] tests/types/bracket.ycp:12 Expected 'integer', seen 'string'.
Parsed:
----------------------------------------------------------------------
Parser error
----------------------------------------------------------------------
[Interpreter] tests/types/bracket.ycp:10 Parser error
Parsed:
----------------------------------------------------------------------
{
    // list <string> a
    // string res
    // filename: "tests/types/bracket.ycp"
    list <string> a = ["foo"];
    string res = a[0]:nil;
    return (res == "foo");
}
----------------------------------------------------------------------
Parsed:
----------------------------------------------------------------------
{
    // list <string> a
    // string res
    // filename: "tests/types/bracket.ycp"
    list <string> a = ["foo"];
    string res = a[0]:"";
    return (res == "foo");
}
----------------------------------------------------------------------
[Parser] tests/types/bracket.ycp:32 Bracket default to 'a' has wrong type
[Parser] tests/types/bracket.ycp:32 Expected 'string', seen 'const boolean'.
Parsed:
----------------------------------------------------------------------
Parser error
----------------------------------------------------------------------
[Interpreter] tests/types/bracket.ycp:30 Parser error
Parsed:
----------------------------------------------------------------------
{
    // list <map <string, boolean>> a
    // boolean res
    // filename: "tests/types/bracket.ycp"
    list <map <string, boolean>> a = [$["key":true]];
    boolean res = a[0, "key"]:nil;
    return (res == true);
}
----------------------------------------------------------------------
Parsed:
----------------------------------------------------------------------
{
    // list <map <string, boolean>> a
    // boolean res
    // filename: "tests/types/bracket.ycp"
    list <map <string, boolean>> a = [$["key":true]];
    boolean res = a[0, "key"]:false;
    return (res == true);
}
----------------------------------------------------------------------
[Parser] tests/types/bracket.ycp:54 Bracket default to 'a' has wrong type
[Parser] tests/types/bracket.ycp:54 Expected 'boolean', seen 'string'.
Parsed:
----------------------------------------------------------------------
Parser error
----------------------------------------------------------------------
[Interpreter] tests/types/bracket.ycp:51 Parser error
[Parser] tests/types/bracket.ycp:61 Bracket parameter '0' to 'a' has wrong type
[Parser] tests/types/bracket.ycp:61 Expected 'string', seen 'integer'.
Parsed:
----------------------------------------------------------------------
Parser error
----------------------------------------------------------------------
[Interpreter] tests/types/bracket.ycp:58 Parser error
Parsed:
----------------------------------------------------------------------
{
    // list <map> a
    // any res
    // filename: "tests/types/bracket.ycp"
    list <map> a = [$["key":true]];
    any res = a[0, 0]:nil;
    return (res == nil);
}
----------------------------------------------------------------------
Parsed:
----------------------------------------------------------------------
{
    // list <map> a
    // boolean res
    // filename: "tests/types/bracket.ycp"
    list <map> a = [$["key":true]];
    boolean res = /* any -> const boolean */a[0, 0]:false;
    return (res == false);
}
----------------------------------------------------------------------
Parsed:
----------------------------------------------------------------------
{
    // list <map> a
    // boolean res
    // filename: "tests/types/bracket.ycp"
    list <map> a = [$["key":true]];
    boolean res = /* any -> const boolean */a[0, 0]:false;
    return (res == false);
}
----------------------------------------------------------------------
Parsed:
----------------------------------------------------------------------
{
    // list <map> a
    // map res
    // filename: "tests/types/bracket.ycp"
    list <map> a = [$["key":true]];
    map res = a[0]:nil;
    return (res == $["key":true]);
}
----------------------------------------------------------------------
Parsed:
----------------------------------------------------------------------
{
    // list <term> a
    // integer res
    // filename: "tests/types/bracket.ycp"
    list <term> a = [`t (1, `b)];
    integer res = /* any -> integer */a[0, 0]:-1;
    return (res == 1);
}
----------------------------------------------------------------------
Parsed:
----------------------------------------------------------------------
{
    // list <map> a
    // map res
    // filename: "tests/types/bracket.ycp"
    list <map> a = [$["key":true]];
    map res = a[(0 + 1)]:nil;
    return (res == nil);
}
----------------------------------------------------------------------
Parsed:
----------------------------------------------------------------------
{
    // integer fce (string str)
    // list <map> a
    // map res
    // filename: "tests/types/bracket.ycp"
    integer fce (string str)
{
    return tointeger (str);
}
    list <map> a = [$["key":true]];
    map res = a[fce ("0")]:$[];
    return (res == $["key":true]);
}
----------------------------------------------------------------------
Parsed:
----------------------------------------------------------------------
{
    // list a
    // boolean cond
    // string res
    // filename: "tests/types/bracket.ycp"
    list a = ["foo"];
    boolean cond = true;
    string res = /* any -> string */a[cond ? 0 : 1]:"";
    return (res == "foo");
}
----------------------------------------------------------------------
