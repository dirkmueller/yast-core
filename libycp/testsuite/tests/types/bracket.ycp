# author: visnov@suse.cz
#
# check deep bracket type checking

{
    list a = ["foo"];
    return ( a[0]:nil == "foo" );
}

{
    list a = ["foo"];
    return ( a["bar"]:"no" == "foo" ); // parse error, cannot index using string in a list
}

{
    list<string> a = ["foo"];
    
    string res = a[0]:nil;
    return ( res == "foo" ); 
}

{
    list<string> a = ["foo"];

    string res = a[0]:"";
    return ( res == "foo" ); 
}


{
    list<string> a = ["foo"];
    string res = a[0]:false; // parse error, default value does not match type
    return ( res == "foo" ); 
}


{
    list< map <string, boolean> > a = [ $[ "key" : true ] ];

    boolean res = a[0, "key"]:nil;
    return ( res == true ); 
}

{
    list< map <string, boolean> > a = [ $[ "key" : true ] ];

    boolean res = a[0, "key"]:false;
    return ( res == true ); 
}

{
    list< map <string, boolean> > a = [ $[ "key" : true ] ];

    boolean res = a[0, "key"]:""; // parse error, default value does not match type
    return ( res == true ); 
}

{
    list< map <string, boolean> > a = [ $[ "key" : true ] ];

    boolean res = a[0, 0]:false; // parse error, second index should be string
    return ( res == true ); 
}

{
    list< map > a = [ $[ "key" : true ] ];

    any res = a[0, 0]:nil; // ok, map does not provide types, but runtime error, resulting in "false"
    return ( res == nil ); 
}

{
    list< map > a = [ $[ "key" : true ] ];

    boolean res = a[0, 0]:false; // parse error, map does not provide types and default is NOT used for type determining
    return ( res == false ); 
}


{
    list< map > a = [ $[ "key" : true ] ];

    boolean res = (boolean) (a[0, 0]:false); // ok, map does not provide types, but runtime error, resulting in "false"
    return ( res == false ); 
}


{
    list< map > a = [ $[ "key" : true ] ];

    map res = a[0]:nil; // test a shorter index list
    return res == $[ "key" : true ] ; 
}


{
    list < term > a = [ `t ( 1, `b ) ];
    integer res = (integer) (a[0, 0]:-1);  // ok, for terms we cannot determine types
    return ( res == 1 );
}

{
    list< map > a = [ $[ "key" : true ] ];

    map res = a[0 + 1]:nil; // also test a function as a parameter type
    return res == nil ; 
}

{
    // type of the ternary operator
    list a = [ "foo" ];
    
    boolean cond = true;
    
    string res = a[ cond ? 0 : 1]:"";
    return res == "foo";
}
