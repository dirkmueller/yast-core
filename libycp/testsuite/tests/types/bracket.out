(true)
(nil)
(true)
(true)
(nil)
(true)
(true)
(nil)
(nil)
(true)
(true)
(true)
(true)
(true)
(true)
(true)
